packages/sui-react-router/src/PatternUtils.js

```diff
@@ -0,0 +1,237 @@
+  while ((match = matcher.exec(pattern))) {
+    if (match.index !== lastIndex) {
+      tokens.push(pattern.slice(lastIndex, match.index))
+      regexpSource += escapeRegExp(pattern.slice(lastIndex, match.index))
+    }
+
+    if (match[1]) {
+      regexpSource += '([^/]+)'
+      paramNames.push(match[1])
+    } else if (match[0] === '**') {
+      regexpSource += '(.*)'
+      paramNames.push('splat')
+    } else if (match[0] === '*') {
+      regexpSource += '(.*?)'
+      paramNames.push('splat')
+    } else if (match[0] === '(') {
+      regexpSource += '(?:'
+    } else if (match[0] === ')') {
+      regexpSource += ')?'
+    } else if (match[0] === '\\(') {
+      regexpSource += '\\('
+    } else if (match[0] === '\\)') {
+      regexpSource += '\\)'
+    }
+
+    tokens.push(match[0])
+
+    lastIndex = matcher.lastIndex
+  }
+
+  if (lastIndex !== pattern.length) {
+    tokens.push(pattern.slice(lastIndex, pattern.length))
+    regexpSource += escapeRegExp(pattern.slice(lastIndex, pattern.length))
+  }
+
+  return {
+    pattern,
+    regexpSource,
+    paramNames,
+    tokens
+  }
+}
+
+const CompiledPatternsCache = Object.create(null)
+
+export function compilePattern(pattern) {
+  if (!CompiledPatternsCache[pattern])
+    CompiledPatternsCache[pattern] = _compilePattern(pattern)
+
+  return CompiledPatternsCache[pattern]
+}
+
+/**
+ * Attempts to match a pattern on the given pathname. Patterns may use
+ * the following special characters:
+ *
+ * - :paramName     Matches a URL segment up to the next /, ?, or #. The
+ *                  captured string is considered a "param"
+ * - ()             Wraps a segment of the URL that is optional
+ * - *              Consumes (non-greedy) all characters up to the next
+ *                  character in the pattern, or to the end of the URL if
+ *                  there is none
+ * - **             Consumes (greedy) all characters up to the next character
+ *                  in the pattern, or to the end of the URL if there is none
+ *
+ * The function calls callback(error, matched) when finished.
+ * The return value is an object with the following properties:
+ *
+ * - remainingPathname
+ * - paramNames
+ * - paramValues
+ */
+export function matchPattern(pattern, pathname) {
+  // Ensure pattern starts with leading slash for consistency with pathname.
+  if (pattern.charAt(0) !== '/') {
+    pattern = `/${pattern}`
+  }
+  let {regexpSource, paramNames, tokens} = compilePattern(pattern)
+
+  if (pattern.charAt(pattern.length - 1) !== '/') {
+    regexpSource += '/?' // Allow optional path separator at end.
+  }
+
+  // Special-case patterns like '*' for catch-all routes.
+  if (tokens[tokens.length - 1] === '*') {
+    regexpSource += '$'
+  }
+
+  const match = pathname.match(new RegExp(`^${regexpSource}`, 'i'))
+  if (match == null) {
+    return null
+  }
+
+  const matchedPath = match[0]
+  let remainingPathname = pathname.substr(matchedPath.length)
```

Comment:
```
Warning: Although String.prototype.substr(…) is not strictly deprecated (as in "removed from the Web standards"),
it is considered a legacy function and should be avoided when possible. It is not part of the core JavaScript
language and may be removed in the future. If at all possible, use the substring() method instead.
```

from the [docu](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr)

anyhow, isn't this the function with the memory leak?

Reply:
Good one!